### DISCUSSION TITLE
GPU solver speed problem

### DISCUSSION THREAD

**Discussion by KagamineLenOffical** [osqp/osqp#555] [GPU Implementation]
Hi, i'm trying to cuosqp to speed up.
Here is the size of my test data:
`problem:  variables n = 541884, constraints m = 95541
          nnz(P) + nnz(A) = 2731413`
Running on osqp with intel MKL takes about 120s. But running on cusoqp takes about 150s. And the gpu usage is quite low. I'm wondering what is the bottleneck for this problem? I think there is very little parallelism for cuosqp on this data. Is there a way to solve these problem?

**Comment by KagamineLenOffical** [comment]
Add iteration times for pcg seems to accleration a lot. But float precision seems to be not enough for this problem. Double precision can get a stable result. What setting should I adjust? For now, float precision seems never converge.

**Comment by imciner2** [comment]
Hi @KagamineLenOffical, sorry for not replying earlier. The performance of the GPU backend compared to MKL is difficult to predict and really depends on your problem and its structure.

I am not surprised to hear that float precision is causing issues and double works. Some problems that might be poorly scaled or be difficult to solve require the extra precision in a double to find the optimum solution.

The settings I would suggest experimenting with are the adaptive rho settings. Changing those (such as enabling/disabling adaptation, changing the adaptation interval, etc.) can have an effect on the convergence speed, and I have experienced cases where a problem would just seem to hang (no convergence to a solution) on the GPU when running with a fixed adaptive rho interval instead of automatic (time-based) adaptation.
